### DISCUSSION TITLE
Creating a better resource picker

### DISCUSSION THREAD

**Discussion by bitbrain** [bitbrain/pandora#72] [Proposals]
## The Problem

The current implementation of the resource selector functionally works:

![Screenshot 2023-08-20 155124](https://github.com/bitbrain/pandora/assets/822035/ab926832-882e-45da-8f29-79c0b346e950)

However, compared to the inbuilt Resource picker in Godot Engine it is lacking features:

- no way to define custom resources, e.g. a custom `AtlasTexture` with some region selected
- no way to have any form of preview of what is currently selected
- no differentiation between types, e.g. everything is a `Resource`

One workaround is by saving any custom resource as a `.tres` file and then loading it with Pandora. While this works fine, the workflow for this is not really scalable.

## Proposal

I propose that similar to the native Godot resource picker, it allows for a similar UI:

![Screenshot 2023-08-20 155600](https://github.com/bitbrain/pandora/assets/822035/4cd85850-9cef-471f-bd66-cd57fa1659c4)

Then e.g. selecting `AtlasTexture` provides a detailed UI:
![Screenshot 2023-08-20 155705](https://github.com/bitbrain/pandora/assets/822035/601964a3-9c1e-428e-83d7-c510fc3ae30a)

The type of resource could be determined by a dedicated property setting where the user can select from a list of resource properties. The only challenge for this is adding the actual UI components. I am not suggesting to build our own UI but actually use the UI coming with Godot out of the box. In the Pandora file, the property could still be stored as a path, and then inside a folder, e.g. `res://pandora/resources` a resource file could be generated (`.tres`) containing the detailed information about the resource.

> 🤔 one challenge is that most of these UI components are not exposed in GDScript but are only available in C++ itself.

Feedback appreciated.

**Comment by imothee** [comment]
I was building a similar system myself (plugin and all) and ran into the same issue on selecting my icon as an atlas texture since all my 2d objects are in spritesheets or animation sheets. 

The hack I made was creating a texture2d property and instantiating a EditorResourcePicker with picker.base_type = "Texture2D" and then dragging the atlas in from a template resource that has one exported "Texture2d". That at least sets the data properly and then all you need to handle is serializing/deserializing it for now. It's a horrible hack but it unblocked me from maintaining my own version of this and porting to Pandora. I'd love to help and have been exploring/digging into proposals on exposing more of this functionality (or worst case building my own in the short term) and would be happy to collab and contribute to this project.

**Comment by TheWalruzz** [comment]
This one is really needed. I'm building a simple RPG system and while Pandora is great for a lot of things, the fact that you cannot create and edit a new resource in the Pandora editor directly is a really big issue.

As an example, I want to be able to assign inventories to all the Characters in the entity database. Creating a separate resource file for each inventory is really tedious and not scalable at all, given that everything else would be handled by Pandora directly. My current workaround is to have inventory template resources that contain initial state of inventory for some characters, which is then duplicated and assigned to a non-Pandora property in Character entity model, so it could be modified as needed. All the changes are serialized alongside the Character entity, and not directly in it. It works, but it's not very elegant.

Other than that, even with some of quirks and bugs remaining, the workflow with Pandora is already great and really saves a lot of time. The only mandatory thing missing to me is proper Resource editing support.

I wonder if exposing internal resource picker to GDScript could be made into a proposal if a custom solution is a no-go.

**Reply by TheWalruzz** [reply]
I've had an idea. What if Resource type info was added to property settings (like proposed), but also a separate button to create that new resource was added next to the input field in property editor? That would then create the Resource in proper Pandora directory (or maybe wherever user configured it) and immediately switch focus in Inspector to that new Resource?
This would circumvent the custom resource picker issue (at least for now) and allow users to edit/view that Resource immediately. Additional lookup button could get added to quickly change that Inspector focus at any time.
Of course, making it work straight in the editor would be the preferred way to go in the long run, but this solution, while hacky, might work for now.

The only thing is, I'm not sure if it is possible to just hijack the inspector focus (but I think it should be, right?)
